MethodTooLargeException in NativeImageAutoFeatureStep.generateFeature

I'm generating a native image which contains a lot of resources (10.000+), mainly because of some imported webjars.

This results in an error during build:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.0.0.Final:build (build) on project test-project: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.deployment.steps.NativeImageAutoFeatureStep#generateFeature threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runner/AutoFeature.beforeAnalysis (Lorg/graalvm/nativeimage/hosted/Feature$BeforeAnalysisAccess;)V
[ERROR]         at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
[ERROR]         at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:489)
[ERROR]         at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:192)
[ERROR]         at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:203)
[ERROR]         at io.quarkus.deployment.steps.NativeImageAutoFeatureStep.generateFeature(NativeImageAutoFeatureStep.java:439)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:834)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

Which happens of course, because the `beforeAnalysis` became to large and hits the method size limit. Which you already forsaw:

https://github.com/quarkusio/quarkus/blob/e4dbda58094aff76eacdaaa25f8009aca86ab12b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java#L105

I've solved this issue by generating additional methods and I will create a merge request for it.

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
@@ -8,6 +8,7 @@ import java.lang.reflect.Executable;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
+import java.nio.charset.StandardCharsets;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -22,6 +23,7 @@ import org.graalvm.nativeimage.hosted.RuntimeReflection;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.GeneratedNativeImageClassBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ForceNonWeakReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.JniRuntimeAccessBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageProxyDefinitionBuildItem;
@@ -45,6 +47,7 @@ import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.gizmo.TryBlock;
 import io.quarkus.runtime.ResourceHelper;
+import io.quarkus.runtime.graal.ResourcesFeature;
 
 public class NativeImageAutoFeatureStep {
 
@@ -75,12 +78,28 @@ public class NativeImageAutoFeatureStep {
     static final String LOCALIZATION_FEATURE = "com.oracle.svm.core.jdk.localization.LocalizationFeature";
 
     @BuildStep
+    GeneratedResourceBuildItem generateNativeResourcesList(List<NativeImageResourceBuildItem> resources,
+            BuildProducer<NativeImageResourcePatternsBuildItem> resourcePatternsBuildItemBuildProducer) {
+        StringBuilder sb = new StringBuilder();
+        for (NativeImageResourceBuildItem i : resources) {
+            for (String r : i.getResources()) {
+                sb.append(r);
+                sb.append("\n");
+            }
+        }
+        //we don't want this file in the final image
+        resourcePatternsBuildItemBuildProducer.produce(NativeImageResourcePatternsBuildItem.builder()
+                .excludePattern(ResourcesFeature.META_INF_QUARKUS_NATIVE_RESOURCES_TXT).build());
+        return new GeneratedResourceBuildItem(ResourcesFeature.META_INF_QUARKUS_NATIVE_RESOURCES_TXT,
+                sb.toString().getBytes(StandardCharsets.UTF_8));
+    }
+
+    @BuildStep
     void generateFeature(BuildProducer<GeneratedNativeImageClassBuildItem> nativeImageClass,
             List<RuntimeInitializedClassBuildItem> runtimeInitializedClassBuildItems,
             List<RuntimeInitializedPackageBuildItem> runtimeInitializedPackageBuildItems,
             List<RuntimeReinitializedClassBuildItem> runtimeReinitializedClassBuildItems,
             List<NativeImageProxyDefinitionBuildItem> proxies,
-            List<NativeImageResourceBuildItem> resources,
             List<NativeImageResourcePatternsBuildItem> resourcePatterns,
             List<NativeImageResourceBundleBuildItem> resourceBundles,
             List<ReflectiveMethodBuildItem> reflectiveMethods,
@@ -192,13 +211,6 @@ public class NativeImageAutoFeatureStep {
                 }
                 overallCatch.invokeInterfaceMethod(ofMethod(DYNAMIC_PROXY_REGISTRY,
                         "addProxyClass", void.class, Class[].class), proxySupport, array);
-            }
-        }
-
-        for (NativeImageResourceBuildItem i : resources) {
-            for (String j : i.getResources()) {
-                overallCatch.invokeStaticMethod(ofMethod(ResourceHelper.class, "registerResources", void.class, String.class),
-                        overallCatch.load(j));
             }
         }
 

--- a/core/runtime/src/main/java/io/quarkus/runtime/graal/ResourcesFeature.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/graal/ResourcesFeature.java
@@ -1,0 +1,32 @@
+package io.quarkus.runtime.graal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+import org.graalvm.nativeimage.hosted.Feature;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+
+import io.quarkus.runtime.ResourceHelper;
+
+@AutomaticFeature
+public class ResourcesFeature implements Feature {
+
+    public static final String META_INF_QUARKUS_NATIVE_RESOURCES_TXT = "META-INF/quarkus-native-resources.txt";
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(
+                getClass().getClassLoader().getResourceAsStream(META_INF_QUARKUS_NATIVE_RESOURCES_TXT)))) {
+            String line;
+            while ((line = in.readLine()) != null) {
+                if (!line.isEmpty()) {
+                    ResourceHelper.registerResources(line);
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/ResourcesITCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/ResourcesITCase.java
@@ -19,6 +19,11 @@ public class ResourcesITCase extends ResourcesTestCase {
                 .get("/resources/test-resources/excluded/unwanted.txt")
                 .then()
                 .statusCode(404);
+
+        RestAssured.when()
+                .get("/resources/META-INF/quarkus-native-resources.txt")
+                .then()
+                .statusCode(404);
     }
 
 }

